### PR TITLE
JDK-8318016: Per-compilation memory ceiling

### PR DESCRIPTION
**Pull request by tstuefe**
This RFE introduces a new compile command to limit memory usage per compilation. It complements and is based upon the JIT memory statistic introduced with https://bugs.openjdk.org/browse/JDK-8317683.

This new memory limit is applied during compilation for either C1 or C2. It has one of two modes:
- *crash* mode (mainly for compiler devs), which aborts with hs-err file when the ceiling is reached, right in the compiler thread at the allocation point that triggered the ceiling
- *stop* mode, which stops the compilation similarly to how hitting the node limit in C2 would stop the compilation.

#### Usage

The compile command takes the form `-XX:CompileCommand:MemLimit,<method>,<memory size>[~<option>]`

The memory size that can be specified as usual. The command also takes an optional trailing option, separated from the memory size by a tilde (odd choice but needed because of https://bugs.openjdk.org/browse/JDK-8318214).

The `<suboption>` can be either `crash` or `stop` and distinguishes between the aforementioned modes. If omitted, the default is "stop".

Since `MemLimit` is based on the JIT statistic, it is implicitly enabled, so there is no need to explicitly enable it with `MemStat`.


Usage examples:


- crash when reaching 100m for a single compilation:
 
```
-XX:CompileCommand='MemLimit,*.*,100m~crash'
```

- limit all methods to a 2m footprint and print a trailing record at VM shutdown

```
-XX:CompileCommand='MemLimit,*.*,2m' -XX:CompileCommand='MemStat,*.*,print'
```

The JIT statistic table (printed either at VM shutdown or via `jcmd <pid> Compiler.memory`) now has a new column "result" that shows if the compilation was completed successfully or had an error or an oom:

```
                              vvvvv                                                                                                                                                                                                                                                
total     NA        RA        result  #nodes  time    type  #rc thread              method
3036016   687328    1757456   oom     2607    0,320   c2    1   0x00007fcf1c181550  java/lang/invoke/StringConcatFactory::<clinit>(()V)

```

#### Implementation notes

The crash variant works immediately.

The stop variant works delayed: If arena memory reaches the limit in the scope of a compilation, we signal the Compile object (`Compile` for c2, `Compilation` for c1) that an oom has occurred.

For C1, we set the bailout string right here. That causes the C1 compilation to fail at the next bailout check.

For C2, we cannot just set `Compile::_failure_reason` since I found that the C2 is not well prepared to bail out from wherever it is tested. So I incorporated the oom check in the nodelimit check C2 does. Note that even that needs some preparatory work (see https://github.com/openjdk/jdk/pull/16205, a prerequisite for this RFE).


Due to this delay, the actual memory used per compilation may be somewhat higher than the limit. In practice the limit works reasonably well, though.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318016](https://bugs.openjdk.org/browse/JDK-8318016): Per-compilation memory ceiling (**Enhancement** - P4)


### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [8e0ab821](https://git.openjdk.org/jdk/pull/16220/files/8e0ab8217a0941f7aebc1020e22a13a542dd523d)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [55e12f68](https://git.openjdk.org/jdk/pull/16220/files/55e12f680744fc90f766d4f37aa83657c4d53353)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16220/head:pull/16220` \
`$ git checkout pull/16220`

Update a local copy of the PR: \
`$ git checkout pull/16220` \
`$ git pull https://git.openjdk.org/jdk.git pull/16220/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16220`

View PR using the GUI difftool: \
`$ git pr show -t 16220`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16220.diff">https://git.openjdk.org/jdk/pull/16220.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16220#issuecomment-1776576409)

### CODE REVIEW

File: src/hotspot/share/compiler/compilationMemoryStatistic.cpp [hunk 1]
```diff
@@ -368,6 +395,20 @@ void CompilationMemoryStatistic::on_end_compilation() {
     arena_stat->print_on(tty);
     tty->cr();
   }
+
+  // Store result
+  // For this to work, we must call on_end_compilatio()n at a point where
```
Comment (rwestrel) [inline src/hotspot/share/compiler/compilationMemoryStatistic.cpp:400]: Typo: on_end_compilatio()n

**Review by rwestrel — APPROVED** [review]
That looks good to me.
Have you done any stress testing? Maybe running compile the world with a low memory limit.

**Comment by tstuefe** [comment]
> That looks good to me. Have you done any stress testing? Maybe running compile the world with a low memory limit.

Many thanks @rwestrel .

I did stress testing by adding a random (low) limit and running GHAs with it. I also ran with -Xcomp -Xbatch. I found a bunch of missing bailout chains, which I took care of before posting this PR (https://bugs.openjdk.org/browse/JDK-8318445, https://bugs.openjdk.org/browse/JDK-8318183). 

But there are more (e.g. https://github.com/openjdk/jdk/pull/16375). We have been discussing a better way to care for these errors (I should have done this on the mailing list, sorry) and maybe introduce a semi-automatic way. Experiments of mine are here: https://github.com/openjdk/jdk/pull/16289 - but I try to find a way that is less invasive.

Finally, there is https://github.com/openjdk/jdk/pull/16247, which adds better diagnostic information to the hs-err file when we miss a bailout.

Bottom line, there are still outstanding issues with missing bailouts, and possibly a better solution at some point. In my view this does not prevent this RFE from going forward.

**Comment by rwestrel** [comment]
> Bottom line, there are still outstanding issues with missing bailouts, and possibly a better solution at some point. In my view this does not prevent this RFE from going forward.

Thanks for the details. That sounds reasonable to me.

File: test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java:2]:
```suggestion
 * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```

File: test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+ * @requires vm.compiler2.enabled
+ * @summary Checks that -XX:CompileCommand=MemLimit,...,crash causes compilation to crash
+ * @library /test/lib
+ * @run driver compiler.print.CompileCommandMemLimit crash true
+ */
+
+/*
+ * @test id=c1_stop
+ * @requires vm.compiler1.enabled
+ * @summary Checks that -XX:CompileCommand=MemLimit,...,crash causes compilation to stop
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java:43]:
```suggestion
 * @summary Checks that -XX:CompileCommand=MemLimit,...,stop causes compilation to stop
```

File: test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+ * @library /test/lib
+ * @run driver compiler.print.CompileCommandMemLimit stop false
+ */
+
+/*
+ * @test id=c2_stop
+ * @requires vm.compiler2.enabled
+ * @summary Checks that -XX:CompileCommand=MemLimit,...,crash causes compilation to stop
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java:51]:
```suggestion
 * @summary Checks that -XX:CompileCommand=MemLimit,...,stop causes compilation to stop
```

**Review by TobiHartmann — APPROVED** [review]
That's a very useful feature, the implementation looks good to me. I've submitted some testing and will report back once it passed.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by tstuefe** [comment]
Many thanks @TobiHartmann!

/integrate

**Pull request closed without merging**